File: src/chess/game/pieces/Knight.java
package game.pieces;

// Import Java modules
import java.awt.Color;
import java.util.ArrayList;
import java.util.Collection;
import game.boardUtils.Square;
import game.utils.Constants;
import game.utils.FileC;
import game.utils.RankC;
import game.utils.UtilMethods;

public class Knight extends Piece {

    private ArrayList<Square> attackers;

    public Knight(FileC file, RankC rank, Color color) {
        super("Knight",file,rank,color);
        attackedPieces = new ArrayList<>();
        attackers = new ArrayList<>();
    }

    public ArrayList<Square> getAttackers() {
        return(attackers);
    }

    @Override
    public ArrayList<Square> getPossiblePositions(String pos, Collection<Square> squares) {
        ArrayList<Square> possiblePositions = new ArrayList<>();
        FileC currFile = FileC.valueOf(pos.substring(0, 1));
        RankC currRank = RankC.valueOf(pos.substring(1));
        Square backLeft1 = this.getSquare(squares, FileC.valueOf(Character.toString((char) (((int) currFile.file())-1))), RankC.valueOf(Integer.toString(currRank.rank()-2)));
        Square backRight1 = this.getSquare(squares, FileC.valueOf(Character.toString((char) (((int) currFile.file())+1))), RankC.valueOf(Integer.toString(currRank.rank()-2)));
        Square backRight2 = this.getSquare(squares, FileC.valueOf(Character.toString((char) (((int) currFile.file())+2))), RankC.valueOf(Integer.toString(currRank.rank()-1)));
        Square forwartRight2 = this.getSquare(squares, FileC.valueOf(Character.toString((char) (((int) currFile.file())+2))), RankC.valueOf(Integer.toString(currRank.rank()+1)));
        Square forwardRight1 = this.getSquare(squares, FileC.valueOf(Character.toString((char) (((int) currFile.file())+1))), RankC.valueOf(Integer.toString(currRank.rank()+2)));
        Square ForwardLeft1 = this.getSquare(squares, FileC.valueOf(Character.toString((char) (((int) currFile.file())-1))), RankC.valueOf(Integer.toString(currRank.rank()+2)));
        Square forwardLeft2 = this.getSquare(squares, FileC.valueOf(Character.toString((char) (((int) currFile.file())-2))), RankC.valueOf(Integer.toString(currRank.rank()+1)));
        Square backLeft2 = this.getSquare(squares, FileC.valueOf(Character.toString((char) (((int) currFile.file())-2))), RankC.valueOf(Integer.toString(currRank.rank()-1)));
        possiblePositions.add(backLeft1);
        possiblePositions.add(backLeft2);
        possiblePositions.add(backRight1);
        possiblePositions.add(backRight2);
        possiblePositions.add(ForwardLeft1);
        possiblePositions.add(forwardLeft2);
        possiblePositions.add(forwardRight1);
        possiblePositions.add(forwartRight2);
        return(possiblePositions);
    }

    @Override
    public boolean validate(FileC f, RankC r, Collection<Square> squares, King king) {
        int[] pos = UtilMethods.getWindowPos(this.getFile(), this.getRank(), Constants.SQUARE_WIDTH, Constants.SQUARE_HEIGHT);
        ArrayList<Square> possiblePositions = this.getPossiblePositions(UtilMethods.getGamePos(pos[0], pos[1], Constants.SQUARE_WIDTH, Constants.SQUARE_HEIGHT), squares);
        Square startingSquare = this.getSquare(squares, this.getFile(), this.getRank());
        Square endSquare = this.getSquare(squares, f, r);
        if (startingSquare != null && endSquare != null) {
            if (endSquare.getPieceColor() != this.getColor()) {
                Piece temp = startingSquare.setPiece(endSquare.getPiece());
                if (!(temp instanceof Empty)) {
                    endSquare.setPiece(new Empty(startingSquare.getRank(), startingSquare.getFile(), temp.getColor()));
                } else {
                    endSquare.setPiece(temp);
                }
                boolean valid = !king.check(king.getFile(), king.getRank(), squares);
                endSquare.setPiece(startingSquare.getPiece());
                startingSquare.setPiece(temp);
                return(valid);
            } else {
                return(false);
            }
        } else {
            return(false);
        }
    }

}

File: bin/chess/game/pieces/knight.py
import pygame
from typing import Sequence, Literal

from board_utils import Square
from chess.CONSTANTS import (SQUARE_WIDTH, WHITE, BLACK)
from .piece import Piece
from .king import King
from .empty import Empty
from utils.flatten import flatten
from utils.types import (
    GamePosition, Positions, Squares
)
from utils.functions import get_game_pos, get_window_pos


class Knight(Piece):

    instances = []

    def __init__(self, file: str, rank: int, color: pygame.Color, min_x: int, max_x: int, min_y: int, max_y: int, square_width: int, square_height: int, win_width: int, win_height: int) -> None:
        pygame.init()
        super().__init__(file, rank, 'Knight', color)
        self.min_x = min_x
        self.max_x = max_x
        self.min_y = min_y
        self.max_y = max_y
        self.square_width = square_width
        self.square_height = square_height
        self.win_width = win_width
        self.win_height = win_height
        self.x, self.y = self.piece_x, self.piece_y = get_window_pos(
            self.file, self.rank, self.possible_files, self.square_width, self.square_height)
        self.attacked_pieces = []

    @classmethod
    def get_possible_positions(self, current_position: GamePosition, squares: Squares) -> Positions:
        file, rank = current_position
        backLeft1 = (self.possible_files[self.possible_files.index(
            file)-1], rank-2) if file != self.possible_files[0] and rank > 2 else (None, 0)
        backRight1 = (self.possible_files[self.possible_files.index(
            file)+1], rank-2) if file != self.possible_files[-1] and rank > 2 else (None, 0)
        backRight2 = (self.possible_files[self.possible_files.index(
            file)+2], rank-1) if file not in self.possible_files[-2:] and rank > 1 else (None, 0)
        forwardRight2 = (self.possible_files[self.possible_files.index(
            file)+2], rank+1) if file not in self.possible_files[-2:] and rank < 8 else (None, 0)
        forwardRight1 = (self.possible_files[self.possible_files.index(
            file)+1], rank+2) if file != self.possible_files[-1] and rank < 7 else (None, 0)
        forwardLeft1 = (self.possible_files[self.possible_files.index(
            file)-1], rank+2) if file != self.possible_files[0] and rank < 7 else (None, 0)
        forwardLeft2 = (self.possible_files[self.possible_files.index(
            file)-2], rank+1) if file != self.possible_files[:2] and rank < 8 else (None, 0)
        backLeft2 = (self.possible_files[self.possible_files.index(
            file)-2], rank-1) if file != self.possible_files[:2] and rank > 1 else (None, 0)
        return list(filter((lambda i: (squares[i[0]+str(i[1])].piece.color != self.color and all(i))), [backLeft1, backLeft2, backRight1, backRight2, forwardRight1, forwardRight2, forwardLeft1, forwardLeft2]))

    def validate(self, position: GamePosition, squares: Squares, king: King) -> bool:
        file, rank = position
        if isinstance(squares[file+str(rank)], Empty):
            squares[file+str(rank)].piece, squares[self.file+str(self.rank)
                                                   ].piece = squares[self.file+str(self.rank)].piece, squares[file+str(rank)].piece
            return not king.check(position=position, squares=squares)
        else:
            if squares[file+str(rank)].piece.color != self.color:
                squares[file+str(rank)
                        ].piece = Empty(tuple(None for _ in range(11)))
            squares[file+str(rank)].piece, squares[self.file+str(self.rank)
                                                   ].piece = squares[self.file+str(self.rank)].piece, squares[file+str(rank)].piece
            return not king.check(position=position, squares=squares)
